Sources/AsyncQueue/MainActorQueue.swift

```diff
@@ -0,0 +1,110 @@
+// MIT License
+//
+// Copyright (c) 2023 Dan Federman
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+/// A queue that enables enqueing ordered asynchronous tasks from a nonisolated context onto the `@MainActor`'s isolated context.
+/// Tasks are guaranteed to begin executing in the order in which they are enqueued. However, if a task suspends it will allow subsequently enqueued tasks to begin executing.
+/// This queue exhibits the execution behavior of an actor: tasks sent to this queue can re-enter the queue, and tasks may execute in non-FIFO order when a task suspends.
+///
+/// A `MainActorQueue` ensures tasks sent from a nonisolated context to the `@MainActor`'s isolated context begin execution in order.
+public final class MainActorQueue: Sendable {
+
+    // MARK: Initialization
+
+    /// Instantiates a main actor queue.
+    public init() {
+        var capturedTaskStreamContinuation: AsyncStream<@Sendable @MainActor () async -> Void>.Continuation? = nil
+        let taskStream = AsyncStream<@Sendable @MainActor () async -> Void> { continuation in
+            capturedTaskStreamContinuation = continuation
+        }
+        // Continuation will be captured during stream creation, so it is safe to force unwrap here.
+        // If this force-unwrap fails, something is fundamentally broken in the Swift runtime.
+        taskStreamContinuation = capturedTaskStreamContinuation!
+
+        Task.detached { @MainActor in
+            for await task in taskStream {
+                await MainActor.shared.suspendUntilStarted(task)
+            }
+        }
+    }
+
+    deinit {
+        taskStreamContinuation.finish()
+    }
+
+    // MARK: Public
+
+    /// An easily accessible instance of a `MainActorQueue`.
+    public static let shared = MainActorQueue()
```

Comment:
I'm tempted to make a free function `MainTask` that takes a `Sendable @MainActor () async -> Void` as an argument and executes the closure on `MainActorQueue.shared`. Basically a `Task { @MainActor` except that it runs on the `MainActorQueue` so you can guarantee ordering.

Reply:
How often would you want to manage multiple `MainActorQueue` instances?

Reply:
I don't think you'd really ever want to manage multiple instances, since ordering between instances is not guaranteed which could lead to confusion.

Reply:
Given the above I just made the `init` for `MainActorQueue` internal: 1d4842d

Creating a `MainTask` type is out of scope – I just spiked on it and it's currently more effort than it's worth.